tools/api-client/python/lib/random_ai.py

```diff
@@ -1405,6 +1414,24 @@ def _should_allow_fire_cancellation(self):
     if self.decision_state['cancelled_fire_attacks'] == 0: return True
     return random.random() < (float(1) / (self.decision_state['cancelled_fire_attacks']))
 
+  def _all_known_attack_types(self):
+    funcPrefix = '_game_action_start_turn_find_attack_'
+    return [ x[len(funcPrefix):] for x in sorted(dir(self)) if x.startswith(funcPrefix) ]
+
+  # Given the list of valid attack types, choose one
+  # With low probability, choose an invalid attack type, and expect it to fail.
+  def _choose_attack_type(self, validAttackTypes):
+    invalidAttackTypes = self._all_known_attack_types()
+    if random.random() < 0.01:
+      invalidAttackType = self._random_array_element(invalidAttackTypes)
+      if invalidAttackType not in validAttackTypes:
```

Comment:
So, I get that for some low probability (less than 1%), we'll get an invalid attack type here. However, I think that the code (or perhaps the variable names) are somewhat inaccurate.

invalidAttackTypes never actually ends up holding an array of invalid attack types. Nor is invalidAttackType guaranteed to hold an invalid attack type here.

If I were to rewrite this code to match the variable names, I'd expect to have to do a set subtraction of valid attack types from all attack types.

Reply:
Ah, you're right, i just need to change the variable name to `allAttackTypes` to make it match the logic (or do something like your suggestion, but that's probably more elegance than i need for this).  Good catch.